File: server/app/pensaer/dsl/ast.py
"""Abstract Syntax Tree nodes for the Pensaer DSL.

This module defines all AST node types that represent parsed DSL commands.
Each node type corresponds to a grammar rule and can be executed against
the MCP tool servers.
"""

from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Any


# =============================================================================
# Enums for typed values
# =============================================================================


class WallType(Enum):
    """Wall construction types."""

    BASIC = "basic"
    STRUCTURAL = "structural"
    CURTAIN = "curtain"
    RETAINING = "retaining"


class DoorType(Enum):
    """Door types."""

    SINGLE = "single"
    DOUBLE = "double"
    SLIDING = "sliding"
    FOLDING = "folding"
    REVOLVING = "revolving"
    POCKET = "pocket"


class WindowType(Enum):
    """Window types."""

    FIXED = "fixed"
    CASEMENT = "casement"
    DOUBLE_HUNG = "double_hung"
    SLIDING = "sliding"
    AWNING = "awning"
    HOPPER = "hopper"
    PIVOT = "pivot"


class SwingDirection(Enum):
    """Door swing directions."""

    LEFT = "left"
    RIGHT = "right"
    BOTH = "both"
    NONE = "none"


class VariableRef(Enum):
    """Variable references for element selection."""

    LAST = "$last"
    SELECTED = "$selected"
    WALL = "$wall"


# =============================================================================
# Coordinate Types
# =============================================================================


@dataclass(frozen=True, slots=True)
class Point2D:
    """2D point in model space (meters)."""

    x: float
    y: float

    def as_tuple(self) -> tuple[float, float]:
        return (self.x, self.y)


@dataclass(frozen=True, slots=True)
class Point3D:
    """3D point in model space (meters)."""

    x: float
    y: float
    z: float

    def as_tuple(self) -> tuple[float, float, float]:
        return (self.x, self.y, self.z)


# =============================================================================
# Element References
# =============================================================================


@dataclass(frozen=True, slots=True)
class ElementRef:
    """Reference to an element by UUID or variable.

    Either uuid or variable will be set, not both.
    """

    uuid: str | None = None
    variable: VariableRef | None = None

    def __post_init__(self) -> None:
        if self.uuid is None and self.variable is None:
            raise ValueError("ElementRef must have either uuid or variable")
        if self.uuid is not None and self.variable is not None:
            raise ValueError("ElementRef cannot have both uuid and variable")

    @classmethod
    def from_uuid(cls, uuid: str) -> "ElementRef":
        return cls(uuid=uuid)

    @classmethod
    def from_variable(cls, var: VariableRef) -> "ElementRef":
        return cls(variable=var)

    def is_variable(self) -> bool:
        return self.variable is not None


# =============================================================================
# Base Command
# =============================================================================


@dataclass(kw_only=True)
class Command:
    """Base class for all DSL commands."""

    line: int = 0
    column: int = 0

    def to_mcp_args(self) -> dict[str, Any]:
        """Convert command to MCP tool arguments."""
        raise NotImplementedError


# =============================================================================
# Wall Commands
# =============================================================================


@dataclass
class CreateWallCommand(Command):
    """Create a single wall segment.

    Maps to: geometry-server/create_wall
    """

    start: Point2D
    end: Point2D
    height: float = 3.0
    thickness: float = 0.2
    wall_type: WallType | None = None
    level_id: str | None = None

    def to_mcp_args(self) -> dict[str, Any]:
        args: dict[str, Any] = {
            "start": self.start.as_tuple(),
            "end": self.end.as_tuple(),
            "height": self.height,
            "thickness": self.thickness,
        }
        if self.wall_type:
            args["wall_type"] = self.wall_type.value
        if self.level_id:
            args["level_id"] = self.level_id
        return args


@dataclass
class CreateRectWallsCommand(Command):
    """Create 4 walls forming a rectangle.

    Maps to: geometry-server/create_rectangular_walls
    """

    min_point: Point2D
    max_point: Point2D
    height: float = 3.0
    thickness: float = 0.2

    def to_mcp_args(self) -> dict[str, Any]:
        return {
            "min_point": self.min_point.as_tuple(),
            "max_point": self.max_point.as_tuple(),
            "height": self.height,
            "thickness": self.thickness,
        }


@dataclass
class ModifyWallCommand(Command):
    """Modify an existing wall's property.

    Maps to: geometry-server/modify_parameter
    """

    wall_id: str
    property_name: str
    value: float | str

    def to_mcp_args(self) -> dict[str, Any]:
        return {
            "element_id": self.wall_id,
            "parameter_name": self.property_name,
            "value": self.value,
        }


# =============================================================================
# Door Commands
# =============================================================================


@dataclass
class PlaceDoorCommand(Command):
    """Place a door in a wall.

    Maps to: geometry-server/place_door
    """

    wall_ref: ElementRef
    offset: float
    width: float = 0.9
    height: float = 2.1
    door_type: DoorType | None = None
    swing: SwingDirection | None = None

    def to_mcp_args(self) -> dict[str, Any]:
        args: dict[str, Any] = {
            "offset": self.offset,
            "width": self.width,
            "height": self.height,
        }
        # wall_id will be resolved at execution time if it's a variable
        if self.wall_ref.uuid:
            args["wall_id"] = self.wall_ref.uuid
        if self.door_type:
            args["door_type"] = self.door_type.value
        if self.swing:
            args["swing"] = self.swing.value
        return args


@dataclass
class ModifyDoorCommand(Command):
    """Modify an existing door's property.

    Maps to: geometry-server/modify_parameter
    """

    door_id: str
    property_name: str
    value: float | str

    def to_mcp_args(self) -> dict[str, Any]:
        return {
            "element_id": self.door_id,
            "parameter_name": self.property_name,
            "value": self.value,
        }


# =============================================================================
# Window Commands
# =============================================================================


@dataclass
class PlaceWindowCommand(Command):
    """Place a window in a wall.

    Maps to: geometry-server/place_window
    """

    wall_ref: ElementRef
    offset: float
    width: float = 1.2
    height: float = 1.0
    sill_height: float = 0.9
    window_type: WindowType | None = None

    def to_mcp_args(self) -> dict[str, Any]:
        args: dict[str, Any] = {
            "offset": self.offset,
            "width": self.width,
            "height": self.height,
            "sill_height": self.sill_height,
        }
        if self.wall_ref.uuid:
            args["wall_id"] = self.wall_ref.uuid
        if self.window_type:
            args["window_type"] = self.window_type.value
        return args


@dataclass
class ModifyWindowCommand(Command):
    """Modify an existing window's property.

    Maps to: geometry-server/modify_parameter
    """

    window_id: str
    property_name: str
    value: float | str

    def to_mcp_args(self) -> dict[str, Any]:
        return {
            "element_id": self.window_id,
            "parameter_name": self.property_name,
            "value": self.value,
        }


# =============================================================================
# Opening Command
# =============================================================================


@dataclass
class CreateOpeningCommand(Command):
    """Create a generic opening in a wall.

    Maps to: geometry-server/create_opening
    """

    wall_ref: ElementRef
    offset: float
    width: float
    height: float
    base_height: float = 0.0

    def to_mcp_args(self) -> dict[str, Any]:
        args: dict[str, Any] = {
            "offset": self.offset,
            "width": self.width,
            "height": self.height,
            "base_height": self.base_height,
            "opening_type": "generic",
        }
        if self.wall_ref.uuid:
            args["host_id"] = self.wall_ref.uuid
        return args


# =============================================================================
# Help Command
# =============================================================================


@dataclass
class HelpCommand(Command):
    """Display help information.

    Not an MCP command - handled locally.
    """

    topic: str | None = None

    def to_mcp_args(self) -> dict[str, Any]:
        return {"topic": self.topic}


# =============================================================================
# Parse Result
# =============================================================================


@dataclass
class ParseError:
    """Error encountered during parsing."""

    message: str
    line: int
    column: int
    token_value: str | None = None

    def __str__(self) -> str:
        loc = f"{self.line}:{self.column}"
        if self.token_value:
            return f"Parse error at {loc} near '{self.token_value}': {self.message}"
        return f"Parse error at {loc}: {self.message}"


@dataclass
class ParseResult:
    """Result of parsing a DSL input.

    Contains either a list of commands or a list of errors.
    """

    commands: list[Command] = field(default_factory=list)
    errors: list[ParseError] = field(default_factory=list)

    @property
    def success(self) -> bool:
        return len(self.errors) == 0

    def add_command(self, cmd: Command) -> None:
        self.commands.append(cmd)

    def add_error(self, error: ParseError) -> None:
        self.errors.append(error)

File: app/src/lib/dsl/ast.ts
/**
 * Abstract Syntax Tree nodes for the Pensaer DSL.
 *
 * This module defines all AST node types that represent parsed DSL commands.
 * Each node type corresponds to a grammar rule and can be executed against
 * the MCP tool servers.
 */

// =============================================================================
// Enums for typed values
// =============================================================================

export enum WallType {
  BASIC = "basic",
  STRUCTURAL = "structural",
  CURTAIN = "curtain",
  RETAINING = "retaining",
}

export enum DoorType {
  SINGLE = "single",
  DOUBLE = "double",
  SLIDING = "sliding",
  FOLDING = "folding",
  REVOLVING = "revolving",
  POCKET = "pocket",
}

export enum WindowType {
  FIXED = "fixed",
  CASEMENT = "casement",
  DOUBLE_HUNG = "double_hung",
  SLIDING = "sliding",
  AWNING = "awning",
  HOPPER = "hopper",
  PIVOT = "pivot",
}

export enum SwingDirection {
  LEFT = "left",
  RIGHT = "right",
  BOTH = "both",
  NONE = "none",
}

export enum RoofType {
  FLAT = "flat",
  GABLE = "gable",
  HIP = "hip",
  SHED = "shed",
  MANSARD = "mansard",
}

export enum RoomType {
  BEDROOM = "bedroom",
  BATHROOM = "bathroom",
  KITCHEN = "kitchen",
  LIVING = "living",
  DINING = "dining",
  OFFICE = "office",
  STORAGE = "storage",
  GENERIC = "generic",
}

export enum StairType {
  STRAIGHT = "straight",
  L_SHAPED = "L-shaped",
  U_SHAPED = "U-shaped",
  SPIRAL = "spiral",
  CURVED = "curved",
}

export enum VariableRef {
  LAST = "$last",
  SELECTED = "$selected",
  WALL = "$wall",
}

// =============================================================================
// Coordinate Types
// =============================================================================

export interface Point2D {
  x: number;
  y: number;
}

export interface Point3D {
  x: number;
  y: number;
  z: number;
}

// =============================================================================
// Element References
// =============================================================================

export interface ElementRef {
  uuid?: string;
  variable?: VariableRef;
}

export function createElementRefFromUUID(uuid: string): ElementRef {
  return { uuid };
}

export function createElementRefFromVariable(variable: VariableRef): ElementRef {
  return { variable };
}

// =============================================================================
// Base Command
// =============================================================================

export interface BaseCommand {
  type: string;
  line: number;
  column: number;
}

// =============================================================================
// Wall Commands
// =============================================================================

export interface CreateWallCommand extends BaseCommand {
  type: "CreateWall";
  start: Point2D;
  end: Point2D;
  height: number;
  thickness: number;
  wallType?: WallType;
  levelId?: string;
  material?: string;
}

export interface CreateRectWallsCommand extends BaseCommand {
  type: "CreateRectWalls";
  minPoint: Point2D;
  maxPoint: Point2D;
  height: number;
  thickness: number;
}

export interface ModifyWallCommand extends BaseCommand {
  type: "ModifyWall";
  wallId: string;
  propertyName: string;
  value: number | string;
}

// =============================================================================
// Door Commands
// =============================================================================

export interface PlaceDoorCommand extends BaseCommand {
  type: "PlaceDoor";
  wallRef: ElementRef;
  offset: number;
  width: number;
  height: number;
  doorType?: DoorType;
  swing?: SwingDirection;
}

export interface ModifyDoorCommand extends BaseCommand {
  type: "ModifyDoor";
  doorId: string;
  propertyName: string;
  value: number | string;
}

// =============================================================================
// Window Commands
// =============================================================================

export interface PlaceWindowCommand extends BaseCommand {
  type: "PlaceWindow";
  wallRef: ElementRef;
  offset: number;
  width: number;
  height: number;
  sillHeight: number;
  windowType?: WindowType;
}

export interface ModifyWindowCommand extends BaseCommand {
  type: "ModifyWindow";
  windowId: string;
  propertyName: string;
  value: number | string;
}

// =============================================================================
// Opening Command
// =============================================================================

export interface CreateOpeningCommand extends BaseCommand {
  type: "CreateOpening";
  wallRef: ElementRef;
  offset: number;
  width: number;
  height: number;
  baseHeight: number;
}

// =============================================================================
// Floor Command
// =============================================================================

export interface CreateFloorCommand extends BaseCommand {
  type: "CreateFloor";
  /** Polygon points defining the floor boundary */
  points: Point2D[];
  /** Floor thickness in meters (default 0.15) */
  thickness: number;
  /** Level/elevation (default 0) */
  level: number;
  /** Level ID string (e.g., "Level 1") */
  levelId?: string;
  /** Floor type (slab, deck, etc.) */
  floorType?: string;
}

// =============================================================================
// Roof Command
// =============================================================================

export interface CreateRoofCommand extends BaseCommand {
  type: "CreateRoof";
  /** Polygon points defining the roof boundary */
  points: Point2D[];
  /** Roof type (flat, gable, hip, shed, mansard) */
  roofType: RoofType;
  /** Slope in degrees (default 30) */
  slope: number;
  /** Overhang distance in meters (default 0.5) */
  overhang: number;
  /** Ridge direction (optional, for gable roofs) */
  ridgeDirection?: "x" | "y";
  /** Level ID string */
  levelId?: string;
}

// =============================================================================
// Room Command
// =============================================================================

export interface CreateRoomCommand extends BaseCommand {
  type: "CreateRoom";
  /** Polygon points defining the room boundary */
  points: Point2D[];
  /** Room name (e.g., "Living Room") */
  name?: string;
  /** Room number (e.g., "101") */
  number?: string;
  /** Room type (bedroom, bathroom, kitchen, etc.) */
  roomType?: RoomType;
  /** Room height in meters (default 3.0) */
  height: number;
  /** Level ID string */
  levelId?: string;
}

// =============================================================================
// Stair Command
// =============================================================================

export interface CreateStairCommand extends BaseCommand {
  type: "CreateStair";
  /** Position of the stair base */
  position: Point2D;
  /** Number of risers */
  risers: number;
  /** Riser height in meters */
  riserHeight: number;
  /** Tread depth in meters */
  treadDepth: number;
  /** Stair width in meters */
  stairWidth: number;
  /** Stair type/configuration */
  stairType: StairType;
  /** Level ID string */
  levelId?: string;
}

// =============================================================================
// Help Command
// =============================================================================

export interface HelpCommand extends BaseCommand {
  type: "Help";
  topic?: string;
}

// =============================================================================
// Command Union Type
// =============================================================================

export type Command =
  | CreateWallCommand
  | CreateRectWallsCommand
  | ModifyWallCommand
  | CreateFloorCommand
  | CreateRoofCommand
  | CreateRoomCommand
  | CreateStairCommand
  | PlaceDoorCommand
  | ModifyDoorCommand
  | PlaceWindowCommand
  | ModifyWindowCommand
  | CreateOpeningCommand
  | HelpCommand;

// =============================================================================
// Parse Result
// =============================================================================

export interface ParseError {
  message: string;
  line: number;
  column: number;
  tokenValue?: string;
}

export interface ParseResult {
  commands: Command[];
  errors: ParseError[];
  success: boolean;
}

// =============================================================================
// MCP Conversion Helpers
// =============================================================================

export function commandToMcpArgs(cmd: Command): Record<string, unknown> {
  switch (cmd.type) {
    case "CreateWall":
      return {
        start: [cmd.start.x, cmd.start.y],
        end: [cmd.end.x, cmd.end.y],
        height: cmd.height,
        thickness: cmd.thickness,
        ...(cmd.wallType && { wall_type: cmd.wallType }),
        ...(cmd.levelId && { level_id: cmd.levelId }),
        ...(cmd.material && { material: cmd.material }),
      };

    case "CreateRectWalls":
      return {
        min_point: [cmd.minPoint.x, cmd.minPoint.y],
        max_point: [cmd.maxPoint.x, cmd.maxPoint.y],
        height: cmd.height,
        thickness: cmd.thickness,
      };

    case "ModifyWall":
      return {
        element_id: cmd.wallId,
        parameter_name: cmd.propertyName,
        value: cmd.value,
      };

    case "PlaceDoor":
      return {
        ...(cmd.wallRef.uuid && { wall_id: cmd.wallRef.uuid }),
        offset: cmd.offset,
        width: cmd.width,
        height: cmd.height,
        ...(cmd.doorType && { door_type: cmd.doorType }),
        ...(cmd.swing && { swing: cmd.swing }),
      };

    case "ModifyDoor":
      return {
        element_id: cmd.doorId,
        parameter_name: cmd.propertyName,
        value: cmd.value,
      };

    case "PlaceWindow":
      return {
        ...(cmd.wallRef.uuid && { wall_id: cmd.wallRef.uuid }),
        offset: cmd.offset,
        width: cmd.width,
        height: cmd.height,
        sill_height: cmd.sillHeight,
        ...(cmd.windowType && { window_type: cmd.windowType }),
      };

    case "ModifyWindow":
      return {
        element_id: cmd.windowId,
        parameter_name: cmd.propertyName,
        value: cmd.value,
      };

    case "CreateOpening":
      return {
        ...(cmd.wallRef.uuid && { host_id: cmd.wallRef.uuid }),
        offset: cmd.offset,
        width: cmd.width,
        height: cmd.height,
        base_height: cmd.baseHeight,
        opening_type: "generic",
      };

    case "CreateFloor":
      return {
        points: cmd.points.map((p) => [p.x, p.y]),
        thickness: cmd.thickness,
        level: cmd.level,
        ...(cmd.levelId && { level_id: cmd.levelId }),
        ...(cmd.floorType && { floor_type: cmd.floorType }),
      };

    case "CreateRoof":
      return {
        points: cmd.points.map((p) => [p.x, p.y]),
        roof_type: cmd.roofType,
        slope_degrees: cmd.slope,
        overhang: cmd.overhang,
        ...(cmd.ridgeDirection && { ridge_direction: cmd.ridgeDirection }),
        ...(cmd.levelId && { level_id: cmd.levelId }),
      };

    case "CreateRoom":
      return {
        points: cmd.points.map((p) => [p.x, p.y]),
        ...(cmd.name && { name: cmd.name }),
        ...(cmd.number && { number: cmd.number }),
        ...(cmd.roomType && { room_type: cmd.roomType }),
        height: cmd.height,
        ...(cmd.levelId && { level_id: cmd.levelId }),
      };

    case "CreateStair":
      return {
        position: [cmd.position.x, cmd.position.y],
        risers: cmd.risers,
        riser_height: cmd.riserHeight,
        tread_depth: cmd.treadDepth,
        stair_width: cmd.stairWidth,
        stair_type: cmd.stairType,
        ...(cmd.levelId && { level_id: cmd.levelId }),
      };

    case "Help":
      return { topic: cmd.topic };

    default:
      return {};
  }
}
